src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,273 @@ class io_fragment {
     char* get_current() { return _buf.get_write() + _used_bytes; }
     char* get_write() { return _buf.get_write(); }
 
-    // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes,misc-non-private-member-variables-in-classes)
-    safe_intrusive_list_hook hook;
-
 private:
+    friend class io_fragment_list;
+
+    io_fragment* _next = nullptr;
+    io_fragment* _prev = nullptr;
+
     ss::temporary_buffer<char> _buf;
     size_t _used_bytes;
 };
 
+/**
+ * A double-ly linked (intrusive) list of `io_fragment`.
+ *
+ * We use a hand rolled list instead of boost::intrusive_list to optimize
+ * certain codepaths, namely the move constructor was about 4x faster in the
+ * hand rolled implementation over boost.
+ *
+ * We're also able to enhance the list with features like iterator invalidation
+ * checking and the like.
+ *
+ * Note that this container does not own the memory for the io_fragments it
+ * contains, iobuf is responsible for that.
+ */
+class io_fragment_list {
```

Comment:
nitpick: i might isolate the implementation of `io_fragment_list` to a separate commit, on the off chance we need to revert `iobuf` to `boost::intrusive`.

Reply:
You mean the implementation vs using in iobuf? I'm not sure I see the benefit, if we revert the change from iobuf it will still be in git history and this impl is really only for use in `iobuf`.

Reply:
>implementation vs using in iobuf

yes

> will still be in git history

fair enough!